plugins/inventory/digitalocean.py

```diff
@@ -0,0 +1,186 @@
+plugin: community.digitalocean.digitalocean
+api_token: "{{ api_token }}"
+attributes:
+  - id
+  - name
+  - memory
+  - vcpus
+  - disk
+  - size
+  - image
+  - networks
+  - volume_ids
+  - tags
+  - region
+keyed_groups:
+  - key: do_region.slug
+    prefix: 'region'
+    separator: '_'
+  - key: do_tags | lower
+    prefix: ''
+    separator: ''
+compose:
+  ansible_host: do_networks.v4 | selectattr('type','eq','public')
+    | map(attribute='ip_address') | first
+  class: do_size.description | lower
+  distro: do_image.distribution | lower
+'''
+
+import json
+from ansible.errors import AnsibleParserError
+from ansible.module_utils.urls import Request
+from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'community.digitalocean.digitalocean'
+
+    def verify_file(self, path):
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(
+                    ('do_hosts.yaml', 'do_hosts.yml',
+                     'digitalocean.yaml', 'digitalocean.yml',
+                     'digital_ocean.yaml', 'digital_ocean.yml')
+            ):
+                valid = True
+            else:
+                self.display.vvv(
+                    'Skipping due to inventory source file name mismatch. '
+                    'The file name has to end with one of the following: '
+                    'do_hosts.yaml, do_hosts.yml '
+                    'digitalocean.yaml, digitalocean.yml, '
+                    'digital_ocean.yaml, digital_ocean.yml.')
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        # cache settings
+        self._read_config_data(path)
+        self.cache_key = self.get_cache_key(path)
+
+        self.use_cache = self.get_option('cache') and cache
+        self.update_cache = self.get_option('cache') and not cache
+
+        results = []
+        if not self.update_cache:
+            try:
+                results = self._cache[self.cache_key]['digitalocean']
+            except KeyError:
+                pass
+
+        if not results:
+            if self.cache_key not in self._cache:
+                self._cache[self.cache_key] = {'digitalocean': ''}
+
+        # request parameters
+        base_url = 'https://api.digitalocean.com/v2'
+        endpoint = 'droplets'
+        url = '{0}/{1}'.format(base_url, endpoint)
+
+        api_token = self.get_option('api_token')
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer {0}'.format(api_token)
+        }
+
+        # send request
+        req = Request(headers=headers)
+        try:
+            payload = json.load(req.get(url))
```

Comment:
Is there some potential pagination that needs to be considered?

Reply:
Ugh, yes. https://developers.digitalocean.com/documentation/v2/#links

Reply:
I only tested this with 7 VMs and I got no pagination content back. But we definitely need to call things in a loop (and maybe even set the number of returned instances to 200 to reduce the number of requests).

Reply:
ok, I'll take a look.

Reply:
It seems straightforward, I will implement it soon, but I am not sure if we should raise the limit to 200. Maybe there is a risk of timeout if the response is too big in some environments. The default droplet limit per account is 25 so in most of the cases it doesn't really matter. Or I can introduce a new variable with a default value of 200, so one can tune it if necessary.

Reply:
I changed some variable names to follow the REST API terminology and how it is referred in DigitalOcean documentation. And it was more convenient to to append new value this way.